extended/diffAgainst.ts

```diff
@@ -0,0 +1,298 @@
+import BTree from '../b+tree';
+import { BNode, BNodeInternal, check } from '../b+tree';
+import type { ExtendedTreeInternals } from './shared';
+
+/**
+ * A walkable pointer into a BTree for computing efficient diffs between trees with shared data.
+ * - A cursor points to either a key/value pair (KVP) or a node (which can be either a leaf or an internal node).
+ *   As a consequence, a cursor cannot be created for an empty tree.
+ * - A cursor can be walked forwards using `step`. A cursor can be compared to another cursor to
+ *   determine which is ahead in advancement.
+ * - A cursor is valid only for the tree it was created from, and only until the first edit made to
+ *   that tree since the cursor's creation.
+ * - A cursor contains a key for the current location, which is the maxKey when the cursor points to a node
+ *   and a key corresponding to a value when pointing to a leaf.
+ * - Leaf is only populated if the cursor points to a KVP. If this is the case, levelIndices.length === internalSpine.length + 1
+ *   and levelIndices[levelIndices.length - 1] is the index of the value.
+ */
+type DiffCursor<K, V> = {
+  height: number;
+  internalSpine: BNode<K, V>[][];
+  levelIndices: number[];
+  leaf: BNode<K, V> | undefined;
+  currentKey: K;
+};
+
+/**
+ * Computes the differences between `treeThis` and `treeOther`.
+ * For efficiency, the diff is returned via invocations of supplied handlers.
+ * The computation is optimized for the case in which the two trees have large amounts of shared data
+ * (obtained by calling the `clone` or `with` APIs) and will avoid any iteration of shared state.
+ * The handlers can cause computation to early exit by returning `{ break: R }`.
+ * Neither collection should be mutated during the comparison (inside your callbacks), as this method assumes they remain stable.
+ * @param treeThis The tree whose differences will be reported via the callbacks.
+ * @param treeOther The tree to compute a diff against.
+ * @param onlyThis Callback invoked for all keys only present in `treeThis`.
+ * @param onlyOther Callback invoked for all keys only present in `treeOther`.
+ * @param different Callback invoked for all keys with differing values.
+ */
+export function diffAgainst<K, V, R>(
+  treeThis: BTree<K, V>,
+  treeOther: BTree<K, V>,
+  onlyThis?: (k: K, v: V) => { break?: R } | void,
+  onlyOther?: (k: K, v: V) => { break?: R } | void,
+  different?: (k: K, vThis: V, vOther: V) => { break?: R } | void
+): R | undefined {
+  const thisInternals = treeThis as unknown as ExtendedTreeInternals<K, V>;
```

Comment:
The convention I use for this kind of thing is to define parameters with an awkward suffix like `_`, then remove it on the equivalent internal variable: `const tree1 = tree1_ as unknown as ExtendedTreeInternals<K, V>;` And yeah, given that it's standalone, tree1/tree2 or treeA/treeB seem like good names.

(of course in C# I might do the converse, as param names are stored in the DLL)